Add support for path matcher function (#28)
sindresorhus/find-up: readme.md

	console.log(await findUp(['rainbow.png', 'unicorn.png']));
	//=> '/Users/sindresorhus/unicorn.png'

	console.log(await findUp(dir => Promise.resolve('unicorn.png')));
	//=> '/Users/sindresorhus/unicorn.png'

	console.log(await findUp(dir => dir));

> 	//=> '/Users/sindresorhus'

Reviewer: Can you add more realistic examples? I think one using `fs.existsSync` and custom logic for deciding whether it should be used or not would be good. To give users an indication about the capabilities.

Author: Done. I think I was trying to avoid the noise of importing extra modules in the example. But yeah, that was a silly example.

	console.log(await findUp(['rainbow.png', 'unicorn.png']));
	//=> '/Users/sindresorhus/unicorn.png'

	console.log(await findUp(dir => {

> 		return fs.existsSync(path.join(dir, 'unicorn.png')) && 'foo';

Author: IIRC, I believe the reason I mixed sync and async here is because:

1. As you mentioned, async `fs.exists()` is deprecated.
2. At the time, converting from callbacks to promises would have been some extra noisy code to understand for any async `fs` method.

But now `fs.promises` is stable and we can use it to make a pretty simple example:

```js
const pathExists = filepath => fs.promises.access(filepath).then(_ => true).catch(_ => false);
console.log(await findUp(async (directory) => {
	const isInstalled = await pathExists(path.join(directory, 'node_modules'));
	return isInstalled && 'package.json';
}});
```

What do you think of that?

As for the utility methods, I agree they are useful and I'd also prefer for them to be properties on the `findUp` function. Let's do that. I think we should merge this PR first and then build the utilities on top of this functionality in a separate PR(s), since this is still useful with or without them.

> 		return fs.existsSync(path.join(dir, 'unicorn.png')) && 'foo';

Reviewer: > How about something like this?

👍 

> I think we should merge this PR first and then build the utilities on top of this functionality in a separate PR(s), since this is still useful with or without them.

👍

> 		return fs.existsSync(path.join(dir, 'unicorn.png')) && 'foo';

Reviewer: I realized that if we implement #33, we don't need the file/directory checks, so we can simply implement two utility methods => https://github.com/sindresorhus/find-up/issues/37

#### filename

> Type: `string` `Function`

Reviewer: This is incorrect as it indicates that you can do `findUp([() => {}, () => {}], [options])`. Would be better to document it as another overload with different parameter name.

```md
### findUp(matcher, [options])

...

#### matcher

Type: `Function`
...
```

> Filename of the file to find, or a custom matcher function to be called with each directory until it returns a filepath to stop the search or the root directory has been reached and nothing was found. When using a matcher function, if you want to check whether a file exists, use [`fs.access()`](https://nodejs.org/api/fs.html#fs_fs_access_path_mode_callback) - this is done automatically when `filename` is a string.

Reviewer: I think it would be useful for us to at least expose convenience methods to check whether the path exists in a normalized way. `fs.access()` is not obvious to use and you would have to manually promisify it when using it async.

We already depend on [`path-exists`](https://github.com/sindresorhus/path-exists) in `locate-path`, so we could depend on it and expose it here somehow without any increase in dependencies. I think this makes sense as checking whether the path exist is the most common operation you would do.

Author: I think my documentation made it sound worse than it is. `fs.access()` isn't really necessary. In most use cases for `find-up`, it's sufficient to use `fs.existsSync()` if the filesystem even needs to be checked, and that returns a simple boolean, making it much easier to use here. I've added an example of that to the Usage section instead. Hopefully, a little simpler/clearer now.

I did consider that maybe we should ignore the result of the matcher if the filepath it returns does not exist. But I think it's highly likely that the matcher will be checking the filesystem anyway. And by leaving this up to the matcher, we also enable the user to decide where the search stops in a more deterministic manner (i.e. its decision will never be overridden). I think this is a good foundation to build on top of.

As for convenience methods, are you thinking something like `findUp.exists(matcher)` / `findUp.existsSync(matcher)`? Or maybe a pipeline like `findUp(matcher, findUp.exists)`? My immediate reaction is this merely saves a few characters. If we were to add this, I'd prefer the latter pipeline pattern.

Returns a `Promise` for either the first filepath found (by respecting the order) or `null` if none could be found.

### findUp.sync(filename, [options])
### findUp.sync(matcher, [options])

Returns a filepath or `null`.

### findUp.sync([filenameA, filenameB], [options])

Returns the first filepath found (by respecting the order) or `null` if none could be found.

#### filename

Type: `string`

Filename of the file to find.

#### matcher

Type: `Function`

> A function that will be called with each directory until it returns a filepath to stop the search or the root directory has been reached and nothing was found. Useful if you want to match files with a certain pattern, set of permissions, or other advanced use cases.

Reviewer: ```suggestion
A function that will be called with each directory until it returns a `string` with the path, which stops the search, or the root directory has been reached and nothing was found. Useful if you want to match files with a certain patterns, set of permissions, or other advanced use-cases.
```

> When using async mode, `matcher` may optionally be an `async` function or return a `Promise` for the filepath.

Reviewer: ```suggestion
When using async mode, the `matcher` may optionally be an async or promise-returning function that returns the path.
```

?
